### DISCUSSION TITLE
Adding support to other LoRa boards/device hats

### DISCUSSION THREAD

**Discussion by owendoeswork** [markqvist/Reticulum#36] [Ideas]
I saw that there was support for the author's RNode LoRa device, but would the same Interface work with other LoRa boards?
Like this one, for example: https://www.amazon.com/Frequency-Spectrum-Modulation-Raspberry-Interface/dp/B07VS47RQZ

Would I need to change any of the code from the RNode Interface to use this device?
I saw that the RNode uses a Semtech SX1276 Chip, but the example board uses a SX1262 Chip. Would that change the syntax needed?

Thanks for any replies.

**Comment by 4c3e** [comment]
Hello! @markqvist already has worked on a tool to configure off-the-shelf LoRa into functional RNodes: [rnodeconfigutil](https://github.com/markqvist/rnodeconfigutil).

It currently supports:

- The original RNode from unsigned.io
- Homebrew RNodes based on ATmega1284p boards
- Homebrew RNodes based on ATmega2560 boards
- Homebrew RNodes based on Adafruit Feather ESP32 boards
- Homebrew RNodes based on generic ESP32 boards
- LilyGO T-Beam v1.1 devices
- LilyGO LoRa32 v2.0 devices
- LilyGO LoRa32 v2.1 devices

I've also had success with using it with a Heltec ESP32 LoRa board.
Please feel free to follow up if you have any other questions.

**Comment by markqvist** [comment]
Yes, the easiest route to support new LoRa transceiver boards is to add support to the [RNode Firmware](https://github.com/markqvist/rnode_firmware) and [RNode Config util](https://github.com/markqvist/rnodeconfigutil).

Adding support for the SX1262 family of devices is already on the roadmap, and work on that should start in the not-too-distant future. You can also have a look at [this section of the Reticulum repo README](https://github.com/markqvist/Reticulum#development-roadmap) for a general overview of where the development is heading.

Speaking in broad terms, there is huge potential for adding support for the many hundreds of kinds of transceiver modules available, and it is most definitely the intention to exploit that potential.

It needs to be done in the most optimally ordered, generalized and efficient way, so we don't end up needing to maintain 150+ codebases for different transceiver types ;) That is also why LoRa transceiver support currently exists under the RNode umbrella.

**Comment by samuk** [comment]
As an aside, I'm going to have a go at getting this open hardware shield working: RFM95, RFM95, RFM97, RFM98. https://github.com/reganhimself/LoRa-Hab-Hat

Regan is going to add an open licence to it & post me a couple of the boards. If anyone wants a board to try let me know and I can post one to you. 

Quite interested in doing a completely open hardware thing, perhaps using https://github.com/Machine-Hum/wifiwart

**Reply by OtisByron** [reply]
@samuk I'm quite excited about the Pi Zero hat. I was just looking for a RNode supported Pi Zero LoRa hat recently. I think it will be a great option until the C port allows lower current consumption options.

I'm willing to purchase a hat and pay shipping (I'm in the USA) at some point. I'm not a software engineer, so feel free to prioritize others that may be able to lend more to the firmware development effort. I'm mostly interested in testing the system as a whole.

Is Regan's plan to sell assembled hats?

**Reply by samuk** [reply]
AFAIK it was just a personal project and he doesn't plan to sell them. If they work then I have some discount options at PCBway & similar, so could get some quotes for PCBA & a dropshipping service. Might be easier to drop him a mail directly: http://reganhimself.com/contact/

These will just be bare boards, so soldering the Lora chip will be required.

**Comment by markqvist** [comment]
What transceiver ICs do the HopeRF modules use? As far as I remember, these are also based on Semtech ICs, aren't they?

Either way, I think it would be great to support them in a way that makes them compatible with the on-air protocol of RNodes (which is very simple). That way, all LoRa based boards can actually talk to each other.

I'd probably be willing to add support for these directly, if someone can send me a bunch of the boards (and a couple of Pi Zeroes to) to test with.

**Reply by markqvist** [reply]
Or, I will also be happy to just help out with some guidance for anyone wanting to take a stab at it. There's a few details that are important to get right with these kinds of things for actually getting usable on-air performance.

**Reply by samuk** [reply]
Semtech SX1278  is that going to work? https://dwmzone.com/en/hoperf/669-rfm98w-rfm98-semtech-sx1278-lora-transceiver-module.html

Might be fun to work with the [$20 Banana Pi Zero](https://www.aliexpress.com/item/1005004108824412.html) for Open(ish) hardware: https://wiki.banana-pi.org/Banana_Pi_BPI-M2_ZERO
(You get [Gerbers ](https://drive.google.com/drive/folders/0B4PAo2nW2KfnU3hlVzVSbnpWQlE?resourcekey=0-HviURYcxVipNgdNYqi14KA)but no design files)

**Reply by markqvist** [reply]
That should work totally fine. Basically, the most imporant requirements is that we have access to the Direct IO lines on the transceiver ICs, for interrupt-driven IO, and that we can query RSSI/signal-detect/signal-level registers on the transceivers fast enough over SPI to implement solid Carrier Detection, as to avoid packet collisions. That is basically the most imporant job of the current RNode implementation :)

**Reply by samuk** [reply]
@markqvist 'd be up for sending you a couple of the Banana Pi Zero's & Lora hats. Would you be up for soldering the Lora chip on?

I'll drop you an email..

**Reply by markqvist** [reply]
That would be awesome. I can solder on the chips, no problem. Thanks a lot @samuk!

**Comment by arizonajon** [comment]
Hi all -

Maybe this is the right place to pursue my questions. I have purchased two V2.1 LILYGO TTGO LoRA32 units, board is marked T3_V1.6.1, and would like to get them running. These boards have the little display, a micro SD card holder, and a microUSB port. One button and one slide switch.

For the unsigned Rnodes, I've been working in an RPi environment with Linux to-date, but am I right to assume that I need to start up an Arduino development environment for working on these boards? 

Cheers and 73 - Jon N7UV

**Reply by arizonajon** [reply]
> Hi all -
> 
> Maybe this is the right place to pursue my questions. I have purchased two V2.1 LILYGO TTGO LoRA32 units, board is marked T3_V1.6.1, and would like to get them running. These boards have the little display, a micro SD card holder, and a microUSB port. One button and one slide switch.
> 
> For the unsigned Rnodes, I've been working in an RPi environment with Linux to-date, but am I right to assume that I need to start up an Arduino development environment for working on these boards?
> 
> Cheers and 73 - Jon N7UV

Continuing to read: See that if I use the RPi, with rnodeconf installed, I should be able to run the command
```rnodeconf --autoinstall```
and it will walk me through setting this up.

So I do so, and after all the setup questions, end up getting the following:

> ...
> Serial port     : /dev/ttyACM0
> Device type     : LilyGO LoRa32 v2.1 850 - 950 MHz
> Platform        : ESP32
> Device MCU      : Espressif Systems ESP32
> Firmware file   : rnode_firmware_latest_lora32v21.zip
> 
> Is the above correct? [y/N] y
> [2022-04-21 11:32:27] Downloading latest frimware from GitHub...
> [2022-04-21 11:32:27] Could not download firmware package
> [2022-04-21 11:32:27] The contained exception was: [Errno 13] Permission denied: './update'
> 

Does this indicate that I'm having a communications issue with GitHub?

Cheers and 73 - Jon N7UV

**Reply by arizonajon** [reply]
Something has changed. I suddenly was able to connect to github and flash the device! There was a Raspberry OS update that took place first, don't know if that had anything to do with it. Now to move to the other device.
cheers and 73 - Jon N7UV

**Reply by markqvist** [reply]
Good to hear you got it running. I'll hide this part of the discussion, since it is a bit off-topic for the original topic. New help requests generally deserves their own topic.
